docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+```bash
+$ apiclient set motd="Hello, Bottlerocket!"
+```
+
+The apiserver would load the current value of `motd` and then create a new object based on the inputted string.
+It would then find the extension responsible for `motd` and make a request to determine whether or not the change should proceed:
+
+```bash
+$ motd proto1 set --setting-version v1 --value '"Hello, Bottlerocket!"'  # Values are passed as JSON, hence the quoting
+```
+
+### Extension Naming & Setting Ownership
+Every top-level setting in Bottlerocket's API must be owned by a singular extension, and that extension will share a name with the setting.
+This name will also be shared by the extension's configuration file and binary components.
+This means that whenever the system must interact with a setting (be it via user action on the API, or rendering templates to be used as configuration files), Bottlerocket can identify the setting extension to invoke in order to satisfy that interaction.
+
+As an example, the following suggests the filesystem layout for a settings extension for the `settings.network` settings:
+
+```
+usr
+└── lib
+    └── bottlerocket
+        └── settings
+            ├── config.d
+            │   └── network.toml
+            └── extensions.d
+                └── network -> /usr/bin/network-extension
+```
+
+#### Transactional Writes and Cross-Validation
+Bottlerocket today is missing a mechanism for cross-validation of settings.
+As an example, suppose we want to represent a range of integers in settings by allowing a `setting.min-value` and `setting.max-value` to be set -- it should be possible for settings extensions to set both simultaneously, while also verifying that the state from setting both is valid (`setting.min_value <= setting.max_value`).
+
+Let us cast this requirement in a different light: imagine an OS administrator wishes to restrict what values can be written to an existing setting -- for example, they may wish to only allow settings values to `network.hosts` so long as they include as a subset the entries mandated by their organization.
+To allow this, settings extensions will be capable of registering themselves as a *validator* for any other setting on the host, including settings that they do not own.
+When handling write requests for multiple settings in a single transaction, the `apiserver` will first gather the resulting writes that would occur under each settings extension involved in the writes.
+Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
+Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.
+
+Validations must specify the setting which triggers the validation in their configuration file:
+
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[extension.validates]
+network = "v1"
```

Comment:
Good questions. I'll update the doc, but the answer to both is no.

At the spec/packaging level, you'd need to specify a dependency on settings that are required. We can add a build-time check to ensure that validated versions are supported.

At runtime either of these will result in an error occurring and the setting failing to be set.

Reply:
I've changed my mind about this. I think the first case is still valid. If you attempt to validate a setting which isn't on the system, then that will silently succeed.

This could happen in cases where a vendor creates a settings extension that you are interested in, and it takes on many responsibilities but you don't particularly care for the validation functionality.

A version not being present will cause validation to fail. In this case, the author clearly wanted a validation to occur against settings which are present on the system but it cannot succeed.

Settings should have a `Requires` line in their RPM package specifying their related settings to prevent these issues at build time.